src/main/java/com/laytonsmith/core/functions/ControlFlow.java

```diff
@@ -117,6 +116,26 @@ public Mixed exec(Target t, Environment env, Mixed... args)
 			return new Class[]{CRECastException.class};
 		}
 
+		@Override
+		public Scope linkScope(StaticAnalysis analysis, Scope parentScope,
+				ParseTree ast, Environment env, Set<ConfigCompileException> exceptions) {
+			if(ast.numberOfChildren() < 2) {
+				return parentScope;
+			}
+
+			// Handle condition in parent scope.
+			Scope condScope = analysis.linkScope(parentScope, ast.getChildAt(0), env, exceptions);
+
+			// Handle if and else branches in separate scopes.
+			analysis.linkScope(analysis.createNewScope(condScope), ast.getChildAt(1), env, exceptions);
```

Comment:
I still would really like to make this a bit more inverted. Instead of passing the analysis in and doing the work manually each time, it would be better for each function to report which scopes each branch has, and then the engine do the actual work of calling linkScope. Where absolutely necessary, this can be written like this, but I think there should be another abstraction layer on top of this that builds off of the VariableScope mechanism.

In any case, there are now two different mechanisms in the code base that do effectively the same work, so one of them should definitely be removed.

Reply:
There is a good default implementation for all functions that do not have to do anything with scoping, so those won't be bothered with this. As for scoping functions, I think behavior of how things depend on each other is so different accross them that adding another layer on top won't do any good. Also, when adding something like the analysis as you suggested (replacement of final values), the way scopes have to be linked might diverge even more.

Reply:
Looking through implementations, I realized that `for()`, `forelse()`, `dowhile()` and `while()` could just call the default implementation instead of handling their arguments in order themselves. The remaining implementations are quite divergent.